### PR TITLE
Use xml builder to render xml response

### PR DESCRIPTION
**Pull request by vpereira**
Hi,

To reduce the amount of code on the controllers, I've changed the XML generation to use XML builder

### CODE REVIEW

**Comment by coolo** [comment]
do you have a rough idea what the performance impact is rendering - let's say /source/openSUSE:Factory?

**Comment by vpereira** [comment]
Hi @coolo, in the wekend I had some time to profile it, I wasn't able to find a performance degradation due XML Builder usage.  Could you please help me to interpret the results or suggest any new test?

### server setup

for p in `osc ls openSUSE:Factory`; do osc copypac -t http://127.0.0.1:3000 openSUSE:Factory $p openSUSE:Factory; sleep 10; done

It downloaded 10700 packages in total.

### branch master

first I did ``git checkout master`` and ran the following:

```python -m timeit -n 1 -r 100 -s 'import os'  'os.system("osc -A $OBS api \"/source/openSUSE:Factory\"")'```

got:

**1 loops, best of 100: 856 msec per loop**

with ```ab```:

```

vpereira@kimura:~/obs_test> ab -n 100 -c 1 -T 'application/xml' -A Admin:opensuse http://192.168.0.20:3000/source/openSUSE:Factory
This is ApacheBench, Version 2.3 <$Revision: 1748469 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking 192.168.0.20 (be patient).....done


Server Software:        
Server Hostname:        192.168.0.20
Server Port:            3000

Document Path:          /source/openSUSE:Factory
Document Length:        339492 bytes

Concurrency Level:      1
Time taken for tests:   14.097 seconds
Complete requests:      100
Failed requests:        0
Total transferred:      34011574 bytes
HTML transferred:       33949200 bytes
Requests per second:    7.09 [#/sec] (mean)
Time per request:       140.969 [ms] (mean)
Time per request:       140.969 [ms] (mean, across all concurrent requests)
Transfer rate:          2356.16 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    1   0.1      1       1
Processing:   113  140  64.9    122     563
Waiting:      107  134  64.9    116     557
Total:        114  141  64.9    122     564

Percentage of the requests served within a certain time (ms)
  50%    122
  66%    128
  75%    132
  80%    148
  90%    185
  95%    193
  98%    562
  99%    564
 100%    564 (longest request)

```

### branch xml-builder

then ```docker-compose stop```, ```git checkout xml-builder```, ```docker-compose up```, repeated the test and got:

**1 loops, best of 100: 900 msec per loop**

the same route with ```ab```

```
vpereira@kimura:~/obs_test> ab -n 100 -c 1 -T 'application/xml' -A Admin:opensuse http://192.168.0.20:3000/source/openSUSE:Factory
This is ApacheBench, Version 2.3 <$Revision: 1748469 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking 192.168.0.20 (be patient).....done


Server Software:        
Server Hostname:        192.168.0.20
Server Port:            3000

Document Path:          /source/openSUSE:Factory
Document Length:        339492 bytes

Concurrency Level:      1
Time taken for tests:   13.450 seconds
Complete requests:      100
Failed requests:        0
Total transferred:      34011582 bytes
HTML transferred:       33949200 bytes
Requests per second:    7.44 [#/sec] (mean)
Time per request:       134.497 [ms] (mean)
Time per request:       134.497 [ms] (mean, across all concurrent requests)
Transfer rate:          2469.54 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    1   0.1      1       1
Processing:   117  134  26.3    123     200
Waiting:      110  127  26.1    116     192
Total:        117  134  26.3    124     200

Percentage of the requests served within a certain time (ms)
  50%    124
  66%    126
  75%    128
  80%    130
  90%    192
  95%    195
  98%    200
  99%    200
 100%    200 (longest request)
vpereira@kimura:~/obs_test> 

```

## Log analysis:

for both branches, analysing the log from server, I don't see the any performance issue in the view layer:

#### branch xml-builder

```
[2f9674d1-79ba-44b6-a896-1c0ab9b64670] [62:1158.41] Completed 200 OK in 174ms (Views: 0.1ms | ActiveRecord: 7.4ms | Backend: 0.0ms | XML: 0.0ms)
[9b8734eb-44f6-4e9a-bbf9-8a1d9352267e] [62:1158.52] Completed 200 OK in 103ms (Views: 0.1ms | ActiveRecord: 8.1ms | Backend: 0.0ms | XML: 0.0ms)
[9dc29e71-d7c3-42df-9e13-ec80f26c3292] [59:1165.29] Completed 200 OK in 107ms (Views: 0.1ms | ActiveRecord: 8.8ms | Backend: 0.0ms | XML: 0.0ms)
[e3922b1d-35e6-48bf-9a93-e67402850dd6] [59:1165.40] Completed 200 OK in 100ms (Views: 0.1ms | ActiveRecord: 7.7ms | Backend: 0.0ms | XML: 0.0ms)
[fea2215e-115a-4b57-a24d-e269c8a4b800] [59:1165.52] Completed 200 OK in 101ms (Views: 0.1ms | ActiveRecord: 7.7ms | Backend: 0.0ms | XML: 0.0ms)
[86c0906b-efd8-4357-a58d-28e952cd3f9c] [62:1159.00] Completed 200 OK in 110ms (Views: 0.1ms | ActiveRecord: 9.5ms | Backend: 0.0ms | XML: 0.0ms)
[8b794c1b-3abd-46ee-98c4-22b290b2c1e9] [62:1159.12] Completed 200 OK in 102ms (Views: 0.1ms | ActiveRecord: 7.1ms | Backend: 0.0ms | XML: 0.0ms)
[2a66ad8d-2b29-471b-bca5-62c22ea41de0] [59:1165.88] Completed 200 OK in 102ms (Views: 0.1ms | ActiveRecord: 7.1ms | Backend: 0.0ms | XML: 0.0ms)
[6994ae4a-dbbb-4f0e-9a8d-8c11f0ebacc5] [52:1157.58] Completed 200 OK in 107ms (Views: 0.1ms | ActiveRecord: 7.3ms | Backend: 0.0ms | XML: 0.0ms)
[b5dbbbaf-463f-426a-897c-751a66d64095] [52:1157.70] Completed 200 OK in 108ms (Views: 0.1ms | ActiveRecord: 7.5ms | Backend: 0.0ms | XML: 0.0ms)
[03f55bfb-8fdd-4b8c-9fff-8c49925ad97f] [52:1157.89] Completed 200 OK in 174ms (Views: 0.1ms | ActiveRecord: 7.2ms | Backend: 0.0ms | XML: 0.0ms)
[ce3d9500-2cca-4d71-bbe6-9788ede9fe32] [52:1158.01] Completed 200 OK in 102ms (Views: 0.1ms | ActiveRecord: 7.2ms | Backend: 0.0ms | XML: 0.0ms)
[89159a75-d4b9-44c3-ba35-9f11fa3596ae] [52:1158.13] Completed 200 OK in 106ms (Views: 0.1ms | ActiveRecord: 6.9ms | Backend: 0.0ms | XML: 0.0ms)
[87a83894-8985-487b-86d4-6f43b207061f] [52:1158.25] Completed 200 OK in 104ms (Views: 0.1ms | ActiveRecord: 9.5ms | Backend: 0.0ms | XML: 0.0ms)
[1cfa5726-928a-434c-9495-fa136bdd58c9] [52:1158.37] Completed 200 OK in 107ms (Views: 0.1ms | ActiveRecord: 8.3ms | Backend: 0.0ms | XML: 0.0ms)
[9296433e-6a23-4496-84b2-d110118ea1a6] [52:1158.50] Completed 200 OK in 110ms (Views: 0.1ms | ActiveRecord: 7.5ms | Backend: 0.0ms | XML: 0.0ms)
[893c0593-8cba-4d4b-ad10-effc694a674d] [52:1158.69] Completed 200 OK in 173ms (Views: 0.1ms | ActiveRecord: 7.5ms | Backend: 0.0ms | XML: 0.0ms)
[3b015baf-0af1-4096-9bb6-26910b8106ca] [52:1158.80] Completed 200 OK in 102ms (Views: 0.1ms | ActiveRecord: 6.9ms | Backend: 0.0ms | XML: 0.0ms)
[adeba75a-8acb-41da-915f-431d18c8b53e] [52:1158.92] Completed 200 OK in 107ms (Views: 0.1ms | ActiveRecord: 7.8ms | Backend: 0.0ms | XML: 0.0ms)
[90e1ffc3-f6e6-41d4-a24e-8c1f79efa798] [52:1159.04] Completed 200 OK in 102ms (Views: 0.1ms | ActiveRecord: 7.2ms | Backend: 0.0ms | XML: 0.0ms)
[7b2c4cca-0e80-4985-a449-4cb63b4c855a] [52:1159.16] Completed 200 OK in 102ms (Views: 0.1ms | ActiveRecord: 7.9ms | Backend: 0.0ms | XML: 0.0ms)
```

#### branch master

```
[a31cd724-45af-42e2-b7ae-a7d5392b3263] [47:6.62] Completed 200 OK in 99ms (Views: 0.1ms | ActiveRecord: 8.3ms | Backend: 0.0ms | XML: 0.0ms)
[1ff29e21-f76f-4cf3-a1a1-16ec2ea746a9] [47:6.74] Completed 200 OK in 104ms (Views: 0.1ms | ActiveRecord: 9.1ms | Backend: 0.0ms | XML: 0.0ms)
[c76ecdeb-8a14-4934-95da-8f99bf911962] [47:6.85] Completed 200 OK in 98ms (Views: 0.1ms | ActiveRecord: 7.2ms | Backend: 0.0ms | XML: 0.0ms)
[ac34a800-8caa-4b3f-98d7-acf98242144c] [47:6.96] Completed 200 OK in 100ms (Views: 0.1ms | ActiveRecord: 7.7ms | Backend: 0.0ms | XML: 0.0ms)
[e0435ab3-34f2-46a0-8962-47eb53506e86] [52:18.31] Completed 200 OK in 98ms (Views: 0.1ms | ActiveRecord: 6.9ms | Backend: 0.0ms | XML: 0.0ms)
[2387ada8-3b96-4b1b-bffb-c11e686d855e] [52:18.43] Completed 200 OK in 95ms (Views: 0.1ms | ActiveRecord: 7.3ms | Backend: 0.0ms | XML: 0.0ms)
[8681d426-7060-4d0f-9aa1-906f739eb787] [52:18.54] Completed 200 OK in 98ms (Views: 0.1ms | ActiveRecord: 7.6ms | Backend: 0.0ms | XML: 0.0ms)
[25ea055f-97cc-464f-beb5-f0751a27af24] [55:19.08] Completed 200 OK in 98ms (Views: 0.1ms | ActiveRecord: 7.6ms | Backend: 0.0ms | XML: 0.0ms)
[2a81bbd0-f928-4bfe-ad32-36d6555f2f7b] [55:19.20] Completed 200 OK in 104ms (Views: 0.1ms | ActiveRecord: 7.3ms | Backend: 0.0ms | XML: 0.0ms)
[12f5ad79-6896-4705-bba8-6bc4d007d18d] [58:6.86] Completed 200 OK in 124ms (Views: 0.1ms | ActiveRecord: 10.8ms | Backend: 0.0ms | XML: 0.0ms)
[f702cb8d-a0de-4bc4-8d60-cae2e13c21fc] [58:6.98] Completed 200 OK in 103ms (Views: 0.1ms | ActiveRecord: 7.4ms | Backend: 0.0ms | XML: 0.0ms)
[17d52e76-a623-45ec-9dc6-deebbdf4c81a] [58:7.10] Completed 200 OK in 98ms (Views: 0.1ms | ActiveRecord: 6.9ms | Backend: 0.0ms | XML: 0.0ms)
[da60b6f0-369a-45bd-8f65-f46c0fa1915a] [55:19.69] Completed 200 OK in 96ms (Views: 0.1ms | ActiveRecord: 7.1ms | Backend: 0.0ms | XML: 0.0ms)
[571b2587-27f4-47d4-a410-9af2e7ae2515] [55:19.80] Completed 200 OK in 97ms (Views: 0.1ms | ActiveRecord: 7.7ms | Backend: 0.0ms | XML: 0.0ms)
[b6ccbb1d-b757-4998-bc60-a0d0f1a77809] [55:19.92] Completed 200 OK in 104ms (Views: 0.1ms | ActiveRecord: 7.7ms | Backend: 0.0ms | XML: 0.0ms)
[b58ccbe7-32ea-432d-b868-10dbbcc431e9] [52:19.59] Completed 200 OK in 94ms (Views: 0.1ms | ActiveRecord: 7.1ms | Backend: 0.0ms | XML: 0.0ms)
[7cc325f7-088e-4b7b-973b-9230593221f0] [52:19.71] Completed 200 OK in 102ms (Views: 0.1ms | ActiveRecord: 8.3ms | Backend: 0.0ms | XML: 0.0ms)
[26ddd7a9-061b-42da-8415-fc4e5dd0b656] [52:19.82] Completed 200 OK in 96ms (Views: 0.1ms | ActiveRecord: 7.0ms | Backend: 0.0ms | XML: 0.0ms)
[2c79bafa-da2b-4f7f-8861-5880ce68c02c] [55:20.37] Completed 200 OK in 99ms (Views: 0.1ms | ActiveRecord: 9.4ms | Backend: 0.0ms | XML: 0.0ms)
[87fc5b8d-43f6-4d29-ad5b-27fe31d73e2f] [55:20.48] Completed 200 OK in 96ms (Views: 0.1ms | ActiveRecord: 7.0ms | Backend: 0.0ms | XML: 0.0ms)
[da2c7907-49bb-4105-be5a-35a0cb65f8ad] [47:8.93] Completed 200 OK in 103ms (Views: 0.1ms | ActiveRecord: 7.1ms | Backend: 0.0ms | XML: 0.0ms)
[f9ee1c64-42d5-4acd-a8db-d2261a23efd9] [55:20.71] Completed 200 OK in 103ms (Views: 0.1ms | ActiveRecord: 6.8ms | Backend: 0.0ms | XML: 0.0ms)
[b3d0e3be-1ee0-4e3f-a1fe-ffe99f2bc09e] [55:20.83] Completed 200 OK in 96ms (Views: 0.1ms | ActiveRecord: 7.4ms | Backend: 0.0ms | XML: 0.0ms)
[8386140d-1eac-4989-a739-648e9e6f245e] [55:20.94] Completed 200 OK in 98ms (Views: 0.1ms | ActiveRecord: 8.0ms | Backend: 0.0ms | XML: 0.0ms)
[55950708-a0dd-43e5-8d35-5a08d56f1022] [55:21.05] Completed 200 OK in 102ms (Views: 0.1ms | ActiveRecord: 6.8ms | Backend: 0.0ms | XML: 0.0ms)
[9e59fc2c-f5aa-45c5-a494-5515892ccbc9] [58:8.69] Completed 200 OK in 102ms (Views: 0.1ms | ActiveRecord: 6.8ms | Backend: 0.0ms | XML: 0.0ms)
[669dcd7d-4591-4846-bd5d-73101aaef27f] [58:8.81] Completed 200 OK in 105ms (Views: 0.1ms | ActiveRecord: 7.3ms | Backend: 0.0ms | XML: 0.0ms)
[7559d6fc-dbc6-47ea-a929-a4612755eaa7] [55:21.40] Completed 200 OK in 99ms (Views: 0.1ms | ActiveRecord: 9.5ms | Backend: 0.0ms | XML: 0.0ms)
[968a59ea-eb8f-4e0e-966d-4b28d83f6d11] [55:21.53] Completed 200 OK in 104ms (Views: 0.1ms | ActiveRecord: 7.4ms | Backend: 0.0ms | XML: 0.0ms)
[93b87dd1-5664-4deb-8b55-f14e605ba72e] [55:21.64] Completed 200 OK in 96ms (Views: 0.1ms | ActiveRecord: 6.6ms | Backend: 0.0ms | XML: 0.0ms)
[63c48b58-aa5e-4dd5-98a1-a81ee5d1492b] [55:21.75] Completed 200 OK in 102ms (Views: 0.1ms | ActiveRecord: 7.3ms | Backend: 0.0ms | XML: 0.0ms)
[43eec19b-8ed7-4d58-8e5b-7aee7cbdb168] [55:21.86] Completed 200 OK in 95ms (Views: 0.1ms | ActiveRecord: 6.8ms | Backend: 0.0ms | XML: 0.0ms)
[a9fb0e8e-2145-4dbb-80af-9b446186f708] [55:21.98] Completed 200 OK in 103ms (Views: 0.1ms | ActiveRecord: 7.1ms | Backend: 0.0ms | XML: 0.0ms)
[9ec9788f-1008-49b4-85cf-78a5a48a25e9] [55:22.13] Completed 200 OK in 134ms (Views: 0.1ms | ActiveRecord: 6.9ms | Backend: 0.0ms | XML: 0.0ms)
[c51b1611-662f-4514-9649-208edf656dab] [58:9.84] Completed 200 OK in 168ms (Views: 0.1ms | ActiveRecord: 7.6ms | Backend: 0.0ms | XML: 0.0ms)
[f20738cb-5cfb-4634-b92f-f5f9051cc230] [58:9.96] Completed 200 OK in 105ms (Views: 0.1ms | ActiveRecord: 9.1ms | Backend: 0.0ms | XML: 0.0ms)
[f125f979-4392-4f34-8edb-4450fe043107] [52:22.11] Completed 200 OK in 95ms (Views: 0.1ms | ActiveRecord: 6.7ms | Backend: 0.0ms | XML: 0.0ms)
[428aa042-8aab-40b1-988e-23a667706abd] [47:11.05] Completed 200 OK in 164ms (Views: 0.1ms | ActiveRecord: 7.2ms | Backend: 0.0ms | XML: 0.0ms)
[2277b6c8-44a8-4805-aa78-c77dc7da5d6e] [47:11.17] Completed 200 OK in 98ms (Views: 0.1ms | ActiveRecord: 7.3ms | Backend: 0.0ms | XML: 0.0ms)
[80b24235-3243-496f-b6fe-8e6d600b799f] [47:11.28] Completed 200 OK in 101ms (Views: 0.1ms | ActiveRecord: 6.7ms | Backend: 0.0ms | XML: 0.0ms)
[97d96008-e576-422c-86ae-0aadcbdf7b68] [47:11.39] Completed 200 OK in 97ms (Views: 0.1ms | ActiveRecord: 6.6ms | Backend: 0.0ms | XML: 0.0ms)
[606a4cfd-1bd2-4503-a8f4-1c11371d7831] [47:11.51] Completed 200 OK in 100ms (Views: 0.1ms | ActiveRecord: 7.1ms | Backend: 0.0ms | XML: 0.0ms)
[416527a0-ebbc-425d-880b-f6d9eb5b21b2] [47:11.63] Completed 200 OK in 104ms (Views: 0.1ms | ActiveRecord: 6.8ms | Backend: 0.0ms | XML: 0.0ms)
[ee0ebe70-ebce-4df2-92b6-f38285b81770] [47:11.81] Completed 200 OK in 165ms (Views: 0.1ms | ActiveRecord: 8.8ms | Backend: 0.0ms | XML: 0.0ms)
[e04c3f26-1ffb-41ac-9e8c-5cec8644e5eb] [47:11.92] Completed 200 OK in 98ms (Views: 0.1ms | ActiveRecord: 6.7ms | Backend: 0.0ms | XML: 0.0ms)
[5dce4c12-8881-4097-8258-4a9b68a92085] [47:12.03] Completed 200 OK in 101ms (Views: 0.1ms | ActiveRecord: 6.5ms | Backend: 0.0ms | XML: 0.0ms)
[e49345ff-0a57-4009-aa17-50e25423eb17] [47:12.15] Completed 200 OK in 97ms (Views: 0.1ms | ActiveRecord: 6.6ms | Backend: 0.0ms | XML: 0.0ms)
[3cff7721-536a-4ec7-bbf0-cea48475f0af] [47:12.26] Completed 200 OK in 99ms (Views: 0.1ms | ActiveRecord: 6.7ms | Backend: 0.0ms | XML: 0.0ms)
[9578a508-ec12-402b-86f2-8b550ecbe390] [47:12.38] Completed 200 OK in 102ms (Views: 0.1ms | ActiveRecord: 7.7ms | Backend: 0.0ms | XML: 0.0ms)
[047862f6-dae4-41dd-9688-1dfb858f320c] [47:12.56] Completed 200 OK in 164ms (Views: 0.1ms | ActiveRecord: 7.5ms | Backend: 0.0ms | XML: 0.0ms)
```

### NOTE

between the branch ```xml-builder``` and ```master``` there is a rails update (rails 5.1.x to 5.2.0) in between that may or may not interfere in the results

**Comment by coolo** [comment]
that difference looks unimportant. your branch looks slower, but master is more volatile. That route isn't as important as it used to be when the webui was an external application. But in 6bebdbe2fcc1c486a3448238caa437fc8df33aa5 I spent quite some time optimizing - but now that I look that was about ?expand=1 - but that shouldn't make a difference with your chances if expand=0 doesn't show a significant difference.

**Comment by vpereira** [comment]
hi @coolo, thank you. 

> that difference looks unimportant. your branch looks slower, but master is more volatile.

looking in the log files, it looks like the difference is in the Activerecord layer and not in the view, so I would assume that the usage of the XML Builders isn't affecting the performance. What do you think? Do you want to see any other route being tested?

**Comment by coolo** [comment]
The difference is not in the active record. the difference there is at max 3ms. The difference is mainly in what is not calculated extra - the controller code. And the reason master is so volatile is because of the ruby memory allocation model with all the strings created and freed and ruby only hitting the GC if some magic number is hit. Your branch always creates a huge XML DOM and then renders it in one go (mostly in C I assume), so you don't have that GC problem.

File: src/api/app/controllers/source_controller.rb [hunk 1]
```diff
@@ -48,16 +48,8 @@ def index
...
-    output = Rails.cache.fetch(['projectlist', Project.maximum(:updated_at), Relationship.forbidden_project_ids]) do
-      dir = Project.pluck(:name).sort
-      output = ''
-      output << "<?xml version='1.0' encoding='UTF-8'?>\n"
-      output << "<directory>\n"
-      output << dir.map { |item| "  <entry name=\"#{::Builder::XChar.encode(item)}\"/>\n" }.join
-      output << "</directory>\n"
-    end
-    render xml: output
+    @projects = Project.pluck(:name).sort
```
Comment (bgeuken) [inline src/api/app/controllers/source_controller.rb]: Please rename @projects to @project_names. That way it's clear that it's a list of names and not objects

File: src/api/app/controllers/source_controller.rb [hunk 2]
```diff
@@ -110,25 +102,12 @@ def show_project
...
-    output = ''
-    output << "<directory count='#{packages.length}'>\n"
-    if params.key? :expand
-      output << packages.map { |p| "  <entry name=\"#{p[0]}\" originproject=\"#{p[1]}\"/>\n" }.join
-    else
-      output << packages.map { |p| "  <entry name=\"#{p}\"/>\n" }.join
-    end
-    output << "</directory>\n"
-    output
+    @packages = params.key?(:expand) ? @project.expand_all_packages : @project.packages.pluck(:name)
```
Comment (bgeuken) [inline src/api/app/controllers/source_controller.rb]: Same here

File: src/api/app/views/source/show_project.xml.builder [hunk 3]
```diff
@@ -0,0 +1,9 @@
+xml.directory(count: @packages.count) do
+  @packages.map do |p|
```
Comment (bgeuken) [inline src/api/app/views/source/show_project.xml.builder]: Please change this to ` @packages.map do |name, project| `

File: src/api/app/views/source/verboseproductlist.xml.builder [hunk 4]
```diff
@@ -1,9 +1,8 @@
-# avoid to render, parser and re-render here, since it can be a hughe content
-
-output = ''
-output << "<productlist count='#{@products.length}'>\n"
-@products.each do |p|
-  output << p.to_axml
+xml.productlist(count: @products.count) do
+  @products.map do |p|
```
Comment (bgeuken) [inline src/api/app/views/source/verboseproductlist.xml.builder]: Please don't use single chars for variable names unless it's a one liner. Naming it `package` makes the code much more readable.

File: src/api/app/controllers/source_controller.rb (hunk 2 above)
Comment (vpereira) [reply src/api/app/controllers/source_controller.rb]: here I think package_names doesnt apply when the parameter ```:expand``` is given..

Comment (bgeuken) [reply src/api/app/controllers/source_controller.rb]: Ah, I see. then keep it that way.

**Comment by bgeuken** [comment]
@vpereira Nice work with the performance analysis 👍

**Pull request merged**